### ISSUE TITLE
Incorrect Momentum Implementation

### ISSUE THREAD

**Issue by eric-tramel** [dfdx/Boltzmann.jl#16]
Previously, a bug existed in `update_weights!` where the learning rate was multiplied against the momentum when calculating the per-mini-batch updates to the RBM weights. This was corrected in the last pull request from @marylou-gabrie.

However, this change has uncovered another bug. I was finding in practice that both CD and PCD failed to converge (or decrease the pseudo-likelihood at all) for non-negligible momentum values (i.e. values above 0.1 or so). After investigating the component values, I found them to be quite off. 

It turns out that the contributions of the *last* weight gradient was not contributing to the *future* weight gradient, as the last weight gradient was only getting added (scaled by momentum of course) to the current set of weights, but was not getting added to gradient itself, which is later copied to the RBM structure for future use.

After making the change I reference below, I was able to get reasonable results. The reason why this issue never appeared before was that the with a small learning rate (i.e. a default value around 0.1) which is further scaled by the training batch size, when the momentum was incorrectly scaled by this term the contribution from this incorrect implementation was negligible and therefore the training errors were not apparent. Only after allowing the momentum contribution to become non-neglible by taking away the incorrect learning rate scaling did the error surface.

The good news is that it was a quick fix.

### Original
```julia
function update_weights!(rbm, h_pos, v_pos, h_neg, v_neg, lr, buf)
    dW = buf
    # dW = (h_pos * v_pos') - (h_neg * v_neg')
    gemm!('N', 'T', 1.0, h_neg, v_neg, 0.0, dW)
    gemm!('N', 'T', 1.0, h_pos, v_pos, -1.0, dW)
    # rbm.W += lr * dW
    axpy!(lr, dW, rbm.W)
    # rbm.W += rbm.momentum * rbm.dW_prev
    axpy!(rbm.momentum, rbm.dW_prev, rbm.W) 
    # save current dW
    copy!(rbm.dW_prev, dW)
end
```

### Modified
```julia
function update_weights!(rbm, h_pos, v_pos, h_neg, v_neg, lr, buf)
    dW = buf
    # dW = (h_pos * v_pos') - (h_neg * v_neg')
    gemm!('N', 'T', lr, h_neg, v_neg, 0.0, dW)
    gemm!('N', 'T', lr, h_pos, v_pos, -1.0, dW)
    # rbm.dW += rbm.momentum * rbm.dW_prev
    axpy!(rbm.momentum, rbm.dW_prev, dW)
    # rbm.W += lr * dW
    axpy!(1.0, dW, rbm.W)
    # save current dW
    copy!(rbm.dW_prev, dW)
end
```

**Comment by dfdx** [comment]
Thanks, this is important issue / fix. 

Since you add so many valuable changes to Boltzmann.jl, I added you as a collaborator to this repository (let me know if someone else from your team needs write access). Though, I would ask you to still make pull requests when making important changes and leave it for 12-24 hours before merging, so that I had a chance to see it.
